Add support for Garmin and Cluetrust Trackpoint extension

gluap opened the pull request:
This pull request adds support for Garmin and Cluetrust (partial) trackpoint extensions.

I started adding Garmin support and then realized that Cluetrust is probably another popular standard, so I added the most prominent features of Cluetrust as well.

Due to the rudimentary nature of the previous trackpoint extension interfaces I changed the interface for accessing the extension attributes as reflected in the changes to the `tests.py`. I am open to suggestions, this is my first pull request and I never wrote XML parsing code before, so feel free to tell me about everything that may be wrong with this commit.

Because of differences in how namespaces were handled in the lxml and minidom based parsers I adapted the minidom based parser in `parser.py` to also discard namespace information for elements during parsing. I did not have to adapt tests in `tests.py` apart from those dealing testing the trackpoint extensions interface.

gluap commented:
Please note that travis has issues setting up the environment lxml in python 3.2 which was fixed by nawagers in https://github.com/tkrajina/gpxpy/pull/101.

nawagers commented:
I've been looking at extending the gpx 1.1 schema for my own purposes (both trackpoints and waypoints). There are many, many extensions to the GPX schema. What I suggest is there is a TrackPointExtension class that holds arbitrary extension data, maybe just as an ElementTree.

The GPXTrackPoint class would hold a list of TrackPointExtension. During parsing, there could be some flag set for identifying extensions (maybe use an xml schema validation?).  If the flag is set, the parsing routine would attempt to identify the extension from a group of known extensions. If the extension is identified, it would create an instance of a derived class from TrackPointExtension instead (like maybe called TPXext_GarminTPv1 or TPXext_Cluetrustgpxdata10).

I think without some level of abstraction between the TrackPointExtension and specific extension types you limit the utility of the library outside of 'known' extensions, where as holding any well-formed and valid data should be a design goal. Thoughts?

joukesmink commented:
Hi Gluap, thanks for sharing this! I am trying to write a Python script that converts GPX files acquired in a half-covered ice rink with bad GPS reception by counting the number of laps and create for each lap a set of GPS coordinates that fits the ice-ring. 
I also want to merge back my heart rate information and was very happy to see your post with support for extgensions. It took a while to download but I think I succeeded. Now I want to test it but I'm still stuck. I tried the following:
        for track in gpx.tracks:
            for segment in track.segments:
                for point in segment.points:
                    p  = point.extensions
                    print (p['TrackPointExtension']['hr'])
But get the error:
TypeError: string indices must be integers

I noticed that you use somewhere mod_gpxpy instead of gpxpy, why is that?

gluap commented:
@nawagers: I think you are right. The way I implemented it was not very extensible. Especially given that there may be any number of extensions people might want to use.

I'm working on re-writing my pull request in a more extensible way taking into account your comments. As I said, I have little experience with either xml or parsers, so your advice was very valuable to me.

I will push one or two intermediate versions to my git branch and comment back here once I consider it done.

@joukesmink: The way I implemented it earlier was not the most optimal solution. If you are using my branch as of https://github.com/tkrajina/gpxpy/pull/102/commits/0bdc7b2a1ddd5c209bc91e8a4ddb23129ec3b467, you will want to access `p.extensions.hr` instead of `p['TrackPointExtension']['hr']`. But I am currently reworking it so my branch may rapidly change.

See the unittests for an example of how I implemented access to `hr`: 
https://github.com/gluap/gpxpy/blob/0bdc7b2a1ddd5c209bc91e8a4ddb23129ec3b467/test.py#L2841

joukesmink commented:
It works! My mistake was that I did not realise that I have to start a new Python consoleI after updating to your version of gpxpy so I was still using the original version.

This works now fine:
                     print (p['TrackPointExtension']['hr'])

And with this:
                    p  = point.extensions
                    for i in p:
                        print(i, p[i])
I get:
TrackPointExtension {'text': '\n      ', 'atemp': '4', 'hr': '139', 'cad': '67'}
TrackPointExtension {'text': '\n      ', 'atemp': '4', 'hr': '140', 'cad': '67'}
TrackPointExtension {'text': '\n      ', 'atemp': '4', 'hr': '142', 'cad': '67'}
TrackPointExtension {'text': '\n      ', 'atemp': '4', 'hr': '144', 'cad': '65'}
TrackPointExtension {'text': '\n      ', 'atemp': '4', 'hr': '145', 'cad': '63'}

Thanks again!!!

nawagers commented:
@gluap I'm no XML expert either, or Python expert for that matter.  After thinking some more on this, maybe additional levels of abstraction isn't the way to go. The trackpoint (and waypoint and routepoint) classes could have a (list of) TPX class that extends ElementTree to add hr, temp, etc. That way, all of the data would be stored in the ElementTree, upon read, and have access to all the standard ElementTree methods for any extension. Then you would have hr and temp attributes, populated during the parse, upon serialization it could put them back in to the ElementTree and just call the super's tostring() to handle the rest. The basic Garmin fields would probably cover 90% of the use cases, just like by Stava: https://strava.github.io/api/v3/uploads/

Should we add any supported schemas to the xsd folder to reference for documentation?

As far as implementation, I would use ElementTree from the standard library. It's been in since 2.5. The cElementTree module has been deprecated. The API should be the same as the lxml etree API for all reading and writing. The problem is that parser options are minidom and lxml. I don't think there is an easy way to use minidom to build an ElementTree. From what I can find online, the ElementTree and cElementTree modules both have better performance than minidom on most XML layouts. Maybe the library should dump minidom in favor of ElementTree completely? I might just be babbling at this point.

tkrajina commented:
Hi all, and  sorry for not being involved in the discussion. I'm very interested in this discussion, but I'm also busy with other projects at the moment :( 

Anyway, this issue touches two problems that are the main sources of complexity in gpxpy. One is that gpxpy can work both with the minidom and lxml parsers (and that means you don't use the "normal" xml parser and dom manipulation, but my own xml "abstraction"). The other problem is that gpxpy uses the same object model for both gpx 1.0 and gpx 1.1. 

Regarding the latter, when I was rewriting gpxpy for 1.1, I wanted to create a way to "describe" the xml<->gpxpy_models transformations, instead of coding it for all fields. I ended up with something like this:

    class GPXTrack:
        gpx_10_fields = [
                mod_gpxfield.GPXField('name'),
                mod_gpxfield.GPXField('comment', 'cmt'),
                mod_gpxfield.GPXField('description', 'desc'),
                mod_gpxfield.GPXField('source', 'src'),
                mod_gpxfield.GPXField('link', 'url'),
                mod_gpxfield.GPXField('link_text', 'urlname'),
                mod_gpxfield.GPXField('number', type=mod_gpxfield.INT_TYPE),
                mod_gpxfield.GPXComplexField('segments', tag='trkseg', classs=GPXTrackSegment, is_list=True),
        ]
        gpx_11_fields = [
                mod_gpxfield.GPXField('name'),
                mod_gpxfield.GPXField('comment', 'cmt'),
                mod_gpxfield.GPXField('description', 'desc'),
                mod_gpxfield.GPXField('source', 'src'),
                'link',
                    mod_gpxfield.GPXField('link', attribute='href'),
                    mod_gpxfield.GPXField('link_text', tag='text'),
                    mod_gpxfield.GPXField('link_type', tag='type'),
                '/link',
                mod_gpxfield.GPXField('number', type=mod_gpxfield.INT_TYPE),
                mod_gpxfield.GPXField('type'),
                mod_gpxfield.GPXExtensionsField('extensions'),
                mod_gpxfield.GPXComplexField('segments', tag='trkseg', classs=GPXTrackSegment, is_list=True),
        ]

https://github.com/tkrajina/gpxpy/blob/master/gpxpy/gpx.py#L1307

Basically, the `gpx_10_fields` array describes how to marshal and unmarshal the xml from/to the object for gpx 1.0. And of course `gpx_11_fields` is the same, but for gpx 1.1. 

Since gpx extension can contain complex xml subtrees, I was thinking that we can use the same "descriptors" to marshal/unmarshal those extension XML subtrees into custom objects. And those objects would use the same descriptors.

For example, a simple GPX extension could be defined like:

    class ExampleGPXExtension:
        __desc__ = [
                mod_gpxfield.GPXField('name'),
                mod_gpxfield.GPXField('comment', 'cmt'),
                mod_gpxfield.GPXField('description', 'desc'),
                mod_gpxfield.GPXField('source', 'src'),
                'link',
                    mod_gpxfield.GPXField('link', attribute='href'),
                '/link',
        ]
        func __init__(self):
            self.name = None
            self.comment = None
            self.description = None
            self.source = None
            self.link = None

Ideas/criticism(s) are welcome :)

nawagers commented:
The API for ElementTree from the standard library is the same as lxml (for what gpxpy uses). I did some testing and imported xml.ElementTree as mod_etree instead of lxml.ElementTree. The standard library version ran and passed all the unit tests and did it about 10% faster than lxml (on python 3.6). Why keep minidom at all? We could easily swap to ElementTree only, just import lxml and fall back on the standard library version.

In regards to extensions, I think part of it comes to the goals of the library. I haven't seen anything explicit, but I suggest these goals of the library:

- Be able to lossless read any well-formed and valid gpx (1.0 or 1.1)
- Be able to manipulate all gpx fields defined by the schema
- Provide convenience functions for common computations and manipulations
- Be able to lossless write out any well-formed and valid gpx

It would be really hard to work with lossless reading and writing if you don't provide a dynamic data structure like an ElementTree. For example, if the library is used for a website that takes in an end user gpx, uses user input to delete a few points, and then provides a new gpx, you don't want the library to modify or delete extension data if the end user isn't expecting that.

By just exposing the data, you allow the developer to handle it if the application needs to interact with the extension fields. This could mean a lot of extra coding to do something simple, like read/write trackpoint.extensions[0].heartrate. I think a good middle ground would be to use an optional wrapper around the ElementTree that can find and modify the most common fields.

gluap commented:
I agree with @nawagers that it depends on the goals and I think the goals he proposes sound worthwile. For me the most important goals from his list are 1 and 4: The library should be idempotent in the sense that it should _allow to losslesly read and write back out any well-formed gpx_.

At first glance, writing the implementation of additional extension classes in the style @tkrajina is suggesting would be the most consistent way of implementing a new extension. If the extension schema is known, it can be implemented and writing it like @tkrajina suggests would allow to write back consistent XML. But one encounters the following problems:

**The tags in extensions live in different namespaces. Currently gpxpy is not namespace-aware.**
Without extensions, caring about namespace was not important because the whole file would be in the GPX namespace. This complicates implementation because different extensions can define the same tags, possibly with different meanings or containing different data types (e.g. cadence as float vs cadence as int). To decide from which namespace a `<cadence>` tag comes and to write it back out into the same namespace, the `GPXfield` class describing it would have to become namespace-aware. So would the parser to use the correct `GPXExtension` for a given element depending on its namespace. This would be required to  _allow to losslesly read and write back out any well-formed gpx_.

**For reading and writing an extension in the spirit of the present API, the schema of any GPX to be read needs to be known beforehand.**
I see no way to losslessly take in the data of yet un-implemented extensions just using the present library design. Without being able to read and write back out unknown extensions, the parser does not  _allow to losslesly read and write back out any well-formed gpx_. Since extensions can use all XML features I agree with @nawagers that using ElementTree (or another library covering XML, ideally from the standard library) would be the way to go. ElementTree has been in the standard library since 2.5 so using it would not pull in external requirements. Storing the whole extension data would therefore be much easier if one could end minidom support (otherwise one would have to convert minidom nodes to ElementTree nodes or vice-versa). 
@tkrajina: are there reasons that @nawagers and I overlooked for keeping minidom, or is it just there for historical reasons? If ElementTree passes the unit tests as a drop-in for lxml for python 2.6 onwards, would minidom still be required?

**Convenience layer for extensions**
I think the most transparent way to implement things is to keep all extension classes as close as possible to the xml schema of the extension itsself and allow direct access to the developer. Therefore I agree with @nawagers, that maybe the most elegant way is to store the whole subtree below the `extensions` in an ElementTree. The convenience functions for reading could then be reduced to a list of common searches in the ElementTree to find the most common types of information, maybe available as @properties of the GPXExtensions object. Possibly the searches could be annotated to return sensible data types (floats for floats, ints for ints, unicode/strings for strings). Writing could stay restricted to direct access to the ElementTree, as for writing developers need to chose between namespaces to write in, the will be forced to find out about their structure anyway. Possibly overwriting existing attributes could still be implemented as a convenience function (by searching and overwriting the content of the existing element in the existing namespace).
If there are design considerations about gpxpy API to take into account (to make working with extensions more alike to tracksegments or trackpoints), one could do that by a wrapper around the ElementTree.

tkrajina commented:
- yes, minidom support is legacy
- if the lxml.ElementTree and ElementTree APIs are the same, that could make some code simpler in https://github.com/tkrajina/gpxpy/blob/master/gpxpy/parser.py (at the moment the XMLParser is a minidom one, and LXMLParser lxml). 

If that works without surprises, then I'm all for implementing your ideas @gluap and @nawagers .

In that case the first thing would be to remove LXMLParser, and rewrite XMLParser to work with a generic ElementTree. 

One thing that worries me is that ElementTree for any element contains a reference to the parent of that node. That means that the gpx object will become much bigger than it is now. It will contain the current object model and the ElementTree of the whole tree, not only extensions. But only the extension element(s) should be changed. 

I like your idea about having a well defined goals for the library. I'll probably add them in the README.

nawagers commented:
I have some working code that completely removes Minidom, sets up Travis to use a matrix for lxml installed vs not, and wrote a unittest to check if lxml is loaded when running in the lxml environment. I want to do some more refactoring relating to this (like numerous docstring references to minidom). I'll submit a PR soon.
